src/ripple/protocol/impl/STAmount.cpp

```diff
@@ -980,6 +980,58 @@ operator- (STAmount const& value)
 // Arithmetic
 //
 //------------------------------------------------------------------------------
+
+// Calculate (a * b) / c when all three values are 64-bit
+// without loss of precision:
+static
+std::uint64_t
+fast_muldiv(
+    std::uint64_t a,
+    std::uint64_t b,
+    std::uint64_t c)
+{
+    boost::multiprecision::uint128_t ret;
+
+    boost::multiprecision::multiply(ret, a, b);
+    ret /= c;
+
+    if (ret > std::numeric_limits<std::uint64_t>::max())
+    {
+        throw std::overflow_error ("overflow: " +
+            std::to_string (a) + " * " +
+            std::to_string (b) + " / " +
+            std::to_string (c));
+    }
+
+    return static_cast<uint64_t>(ret);
+}
+
+static
+std::uint64_t
+fast_muldiv(
```

Comment:
Actually, the name `fast` stuck from when I was testing this implementation, which is faster. I'll come up with a better name.